src/burn/engine/engine.cpp

```diff
@@ -65,6 +65,79 @@ static HRESULT Restart();
 
 // function definitions
 
+extern "C" BOOL EngineInCleanRoom(
+    __in_z_opt LPCWSTR wzCommandLine
+    )
+{
+    static DWORD cchCleanRoomSwitch = lstrlenW(BURN_COMMANDLINE_SWITCH_CLEAN_ROOM);
+
+    // This check is wholly dependent on the clean room command line switch being
+    // present at the beginning of the command line. Since Burn is the only thing
+    // that should be setting this command line option, that is in our control.
+    BOOL fInCleanRoom = (wzCommandLine &&
+        (wzCommandLine[0] == L'-' || wzCommandLine[0] == L'/') &&
+        CSTR_EQUAL == ::CompareStringW(LOCALE_INVARIANT, NORM_IGNORECASE, wzCommandLine + 1, cchCleanRoomSwitch, BURN_COMMANDLINE_SWITCH_CLEAN_ROOM, cchCleanRoomSwitch) &&
+        wzCommandLine[1 + cchCleanRoomSwitch] == L'='
+    );
+
+    return fInCleanRoom;
+}
+
+extern "C" BOOL EngineInPackageCache(
+    __in_z_opt LPCWSTR wzPath
+    )
+{
+    HRESULT hr = S_OK;
+    BOOL fRunningFromCache = FALSE;
+    LPWSTR sczFolder = NULL;
+    BOOL fTryPerMachine = TRUE;
+    GUID guidBundleId = { };
+    WCHAR wzBundleId[39];
+    LPWSTR sczCacheFolder = NULL;
+    int nCompare = 0;
+
+    if (!wzPath || !*wzPath)
+    {
+        ExitFunction();
+    }
+
+    hr = PathGetDirectory(wzPath, &sczFolder);
+    ExitOnFailure(hr, "Failed to get folder from engine path.");
+
+    hr = SectionGetBundleIdFromModule(&guidBundleId);
+    ExitOnFailure(hr, "Failed to get bundle id from module.");
+
+    if (!::StringFromGUID2(guidBundleId, wzBundleId, countof(wzBundleId)))
+    {
+        hr = E_OUTOFMEMORY;
+        ExitOnRootFailure(hr, "Failed to convert bundle guid into string.");
+    }
+
+    for (DWORD i = 0; i < 2; i++)
+    {
+        hr = CacheGetCompletedPath(fTryPerMachine, wzBundleId, &sczCacheFolder);
```

Comment:
Calling into the engine code before `AppInitialize` makes me very uncomfortable.  Could a bundle running in the package cache spawn off a clean room process from the package cache (doesn't copy itself anywhere, just runs itself)?